src/merge.rs

```diff
@@ -322,15 +321,17 @@ pub fn merge(
         }
         // Merge put together the fields of records, and recursively merge
         // fields that are present in both terms
-        (Term::Record(m1, attrs1), Term::Record(m2, attrs2)) => {
+        (Term::Record(mut m1, attrs1), Term::Record(mut m2, attrs2)) => {
             /* Terms inside m1 and m2 may capture variables of resp. env1 and env2.  Morally, we
              * need to store closures, or a merge of closures, inside the resulting record.  We use
              * the same trick as in the evaluation of the operator DynExtend, and replace each such
              * term by a variable bound to an appropriate closure in the environment
              */
             let mut m = HashMap::new();
             let mut env = Environment::new();
-            let (mut left, mut center, mut right) = hashmap::split(m1, m2);
+            rev_thunks(m1.values_mut(), &mut env1);
+            rev_thunks(m2.values_mut(), &mut env2);
```

Comment:
This looks a bit inefficient. There will probably have to be a deep thought on data representation in the abstract machine in the future. But after the release.

Reply:
Do you have a particular aspect in mind? The fact that we are not tracking dependencies of fields and just revert everything blindly, or something else?

Reply:
Indeed, on principle, we don't want to be traversing an environment that we don't know the size of.

That being said, it's probably ok for now. We will just have to make a pass on the execution model with a cooler head. In the future.

Reply:
I'm not sure to follow the "an environment that we don't know the size of". This operation is not dependent on the size of the environment: the environment is just passed to retrieve the thunks and replace them with the reversed thunks (so one `remove` and one `insert` per field). For custom merge function reversing may be more involved, but for recursive records  it is `O(|m1| + |m2|)`, which was already the case for merging without this PR.

What we could do is to only reverse those fields that actually refers to other fields in a recursive manner. I'm not sure it would make a big difference for the operation of merging in itself, but for the evaluation of the program it would avoid invalidating and potentially recomputing thunks that don't really need it.

Reply:
Ah I think I misunderstood. I read `rev_thunks` as reverting all the thunks in the environment. Which it isn't.

As long as the operations are only dependent on the number of fields, I think that it's fair game.